pkg/promtail/targets/file/tailer.go

```diff
@@ -136,33 +162,66 @@ func (t *tailer) markPositionAndSize() error {
 	t.posAndSizeMtx.Lock()
 	defer t.posAndSizeMtx.Unlock()
 
-	pos, err := t.tail.Tell()
+	size, err := t.tail.Size()
 	if err != nil {
+		// If the file no longer exists, no need to save position information
+		if err == os.ErrNotExist {
+			level.Info(t.logger).Log("msg", "skipping update of position for a file which does not currently exist")
+			return nil
+		}
 		return err
 	}
-	readBytes.WithLabelValues(t.path).Set(float64(pos))
-	t.positions.Put(t.path, pos)
+	totalBytes.WithLabelValues(t.path).Set(float64(size))
 
-	size, err := t.tail.Size()
+	pos, err := t.tail.Tell()
 	if err != nil {
 		return err
 	}
-	totalBytes.WithLabelValues(t.path).Set(float64(size))
+	readBytes.WithLabelValues(t.path).Set(float64(pos))
+	t.positions.Put(t.path, pos)
 
 	return nil
 }
 
-func (t *tailer) stop(removed bool) {
+func (t *tailer) stop() {
+	t.stopMtx.Lock()
+	defer t.stopMtx.Unlock()
+
+	// Use some behaviors of go channels to prevent multiple executions of stop.
+	// This is necessary because there are two separate threads which can call this method from filetarget
+	select {
+	case <-t.stopped:
+		// From the spec: A receive operation on a closed channel can always proceed immediately,
+		// yielding the element type's zero value after any previously sent values have been received.
+		return
+	default:
+		// Also from the spec: If one or more of the communications can proceed,
+		// a single one that can proceed is chosen via a uniform pseudo-random selection.
+		// Otherwise, if there is a default case, that case is chosen.
+		// We have no sender on t.stopped so that case can only proceed when the channel is closed.
+		// If we hit the default case it's because nobody has closed the channel and therefore nobody has called stop.
+		break
+	}
```

Comment:
Another way of doing this would be to wrap a call to `stop` with a `sync.Once`, which would also let you get rid of `stopMtx`. It's not clear to me if either is "better" but using the `sync.Once` would probably need less explanation in the code.

Reply:
good feedback, done.